src/useObservableProps.ts

```diff
@@ -0,0 +1,187 @@
+import {
+    action,
+    isObservable,
+    isObservableArray,
+    isObservableMap,
+    isObservableObject,
+    observable,
+    remove,
+    set
+} from "mobx"
+import { useMemo } from "react"
+
+type LocalObservables = WeakMap<any, boolean>
+
+const EMPTY_OBJECT = {}
+
+export type UseObservablePropsMode<P> =
+    | "shallow" // the reference is not changed and the props (primitives, objects, maps and arrays) are turned into a shallowly observable object
+    | "deep" // the reference is not changed and props (primitives, objects, maps and arrays) are turned into a deeply observable object
+    | {
+          deepProps: Array<keyof P> // like 'shallow', except some properties are turned into deep observables 'opt-in'
+      }
+
+export function useObservableProps<P>(props: P, mode: UseObservablePropsMode<P>): P {
```

Comment:
what about something like this?
```
const p = useObservableProps((props: P) => {
  name: props.name,
  deepFoo: props.some.deep.prop,
  someArrayItem: props.someArray[5]
})
```
My question if so though is, what if `.some.deep` is undefined (it throws upon access),
should it:
1. throw and force the user to write `props.some && props.some.deep && props.some.deep.prop`?
2. assume that if it throws the value will be undefined?
3. assume that if it throws the value will become something like `new class UnreachableValue` and test in those cases for `p.deepFoo instanceof UnreachableValue`?

Reply:
Also, what if you pass an array or an object as a property, should it translate those into deep observables, shallow observables or do nothing at all

Reply:
So with your approach, it would mostly just silently ignore that there something missing? :) I mean people should be using prop types and/or static typing to catch this kind of errors. I don't feel like we should do any hand-holding in here. And if they don't use any of those, well then RIP :)

The variant with callback getting props is definitely ok then you can just pass it a selector defined outside of the component which is a good thing for sure.

Reply:
(beginning of rant)
I so wished JS had something like `props.some?.deep?.prop` like c# has...
(end of rant :) )

Ok, so it would be option 1

What about selecting objects/arrays then?
Should those be turned into deep observables? shallow observables?

Or something like:
```
const p1 = useObservableProps(() => ({
  name: props.name,
  someArray: props.array
  someObj: props.obj
}), {
  name: 'ref', // default
  someArray: 'shallow',
  someObj: 'deep'
}); 
```

Reply:
Well. If you pass a simple object, it would be a one time initialized as it is now and then updated from within a component. If it gets the function (eg, selector) it can run it and init/update underlying observable based on changes that selector provided...

Actually, let's keep it as a separate hook, but name it something like `useSyncedObservable`. The point is that it's not really possible to access component's props in the custom hook. It needs to be passed in a second argument which would not make sense if a simple object would be supplied.

```ts
const userSelector = createSelector(props => ({
  firstName: props.firstName,
  lastName: props.lastName
}))

const Component = (props) => {
  const user = useSyncedObservable(userSelector(props))
  // whenever props would change, the user observable would be updated
}
```

Maybe it's kinda overengineered because whenever props change the component would re-render anyway, hmm. However, it might be useful for turning eg. value from context into observable.

Reply:
```js
function useSyncedObservable(data, selector) {
  const obs = useObservable({})
  Object.assign(obs, selector(data))
}

const userSelector = props => ({
  firstName: props.firstName,
  lastName: props.lastName
}))

const user = useSyncedObservable(props, userSelector)
```

Something like that you mean?

Reply:
I would keep the single argument being a result from the selector. No need to be so specific imo. Also, the selector return value is memoried, so it shouldn't be applied every time. The `useEffect` might be good here to run the action it only when the value changes.

I am still not sure about the usefulness of this though. Even turning context into observable might be better done on the context side.

It would need to be more robust also because `useObservable` accepts Array or Map, so to keep it consistent, it would need to have this logic here as well. However, updating arrays is rather tricky. Replace everything, or merge? Someone might have even changed the observable by some effect in the component.

Reply:
> I would keep the single argument being a result from the selector. No need to be so specific imo. Also, the selector return value is memoried, so it shouldn't be applied every time. The useEffect might be good here to run the action only when the value changes.

Hmm? I might be loosing something, but if the mutating variable is not passed, how is the hook going to know what to update to?

The problem with useEffect is that it will run the change -after- the render phase is finished, so the synced observable would always be one step behind the actual values :(
I wish react hooks had any kind of hook that would run -before- the render phase

> It would need to be more robust also because useObservable accepts Array or Map, so to keep it consistent, it would need to have this logic here as well. However, updating arrays is rather tricky. Replace everything, or merge? Someone might have even changed the observable by some effect in the component.

The problem I see with using object assign is that for every "input" change it will create new observable objects for everything that is selected

useObservableProps used deep/shallow object/array/map reconciliation, not sure if you are still interested in using that. It is some kind of "object assign" to observables from plain objects.

Reply:
> Hmm? I might be loosing something, but if the mutating variable is not passed, how is the hook going to know what to update to?

Like I've shown in my example, you would just call it before passing result to the hook. `useSyncedObservable(userSelector(props))`. It would pass the same object if props would stay the same.

> The problem with useEffect is that it will run the change -after- the render phase is finished, so the synced observable would always be one step behind the actual values :(
> I wish react hooks had any kind of hook that would run -before- the render phase

Oh right, keep forgetting about that :/

Ok, let's keep `useObservableProps` then, but it will be completely up to you to respond to bug reports and questions :) I understand half of it and since it's just additive...

Reply:
Hmm, now that I think about it, maybe it can be done just as
```
const C = props => {
  const obs = useObserver({})
  obs.firstName = props.firstName;
  obs.lastName = props.lastName;
  
  const fullName = useComputed(.... using obs.firstName and lastName)
}
```
Makes me wonder if you can do code between hooks though or it will break everything apart

Reply:
In that contrived example of yours I even find easier to just do this...Less verbose in overall, although less intuitive probably.

```ts
const C = props => {
  const fullName = useComputed(/*.... using obs.firstName and lastName*/, [props.firstName, props.lastName])
}
```

Reply:
Yeah, I guess so :)
I hate having to manually add the list of dependencies though, it feels like a step back

Reply:
Well, it's either that or you manually declare observable ;) Obviously, the true power of `useComputed` is when you actually use some observables in there. Otherwise, it's much easier to just have a selector :)

Reply:
Here goes a prophecy: Just wait until people start writing reactions to props stuff and then wonder why it doesn't work :D
Or forget their array in useComputed